fix(test workflow): fix scheduled test to run successfully

### Description

Test action is failing to run with the following output:

```terminal
npm ERR! Cannot read property '@chill-viking/layout' of undefined
```

Possibly an issue with cache buster step that adjusts package-lock.json file manually.

### Steps to Reproduce

```shell
1. Run test workflow
```


### Expected Behavior

Should run all tests successfully.

### Actual Behaviour

Fails on first job with error referenced in description.

### Additional Context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/.github/workflows/run-action.yml
+++ b/.github/workflows/run-action.yml
@@ -21,49 +21,12 @@ on: # yamllint disable-line rule:truthy
         default: "@chill-viking"
 
 jobs:
-  prepare-for-cache:
-    runs-on: ubuntu-latest
-    name: "[prepare]"
-    steps:
-      - name: Download test folder
-        uses: actions/download-artifact@v3
-        with:
-          name: test_package
-          path: testing
-
-      - name: Cache buster
-        shell: pwsh
-        run: |
-          cd testing
-          npm version 1.1.${{ inputs.node_major }}${{ github.run_number }} --no-git-tag-version
-          $lockFile = Get-Content -Path .\package-lock.json -Raw | ConvertFrom-Json -AsHashtable
-          $lockFile.name = "testing action for node ${{ inputs.node_major }}.x on ${{ inputs.runs_on }} #${{ github.run_number }}"
-          $lockFile.version = "1.1.${{ inputs.node_major }}${{ github.run_number }}"
-          $lockFile | ConvertTo-Json -Depth 10 | Set-Content -Path .\package-lock.json
-
-          if (Test-Path node_modules)
-          {
-            Write-Host "::error::node_modules exists, it should not"
-            exit 1
-          }
-          else
-          {
-            Write-Host "node_modules does not exist, able to test action"
-          }
-
-      - name: Upload cache-busted test folder
-        uses: actions/upload-artifact@v3
-        with:
-          name: test_package_node_${{ inputs.node_major }}_${{ inputs.runs_on }}
-          path: testing
-
   pre-cache-run:
     runs-on: ${{ inputs.runs_on }}
-    name: "[pre-cache]"
-    needs: prepare-for-cache
+    name: "initial run"
     outputs:
       duration_ms: ${{ steps.timer-end.outputs.duration_ms }}
-      key_used: ${{ steps.cache-key.outputs.key }}
+      used_cache: ${{ steps.test-run.outputs.restored_from_cache }}
     steps:
       - name: "Use Node.js (${{ inputs.node_major }}.x)"
         uses: actions/setup-node@v3
@@ -73,15 +36,8 @@ jobs:
       - name: Download test folder
         uses: actions/download-artifact@v3
         with:
-          name: test_package_node_${{ inputs.node_major }}_${{ inputs.runs_on }}
+          name: test_package
           path: testing
-
-      - name: Get Cache Key
-        id: cache-key
-        shell: pwsh
-        run: |
-          $key = "${{ inputs.runs_on }}-${{ hashFiles('**/package-lock.json') }}"
-          Write-Output "key=${key}" >> $env:GITHUB_OUTPUT
 
       - name: Download action
         uses: actions/download-artifact@v3
@@ -98,6 +54,7 @@ jobs:
           Write-Output "date=${dateString}" >> $env:GITHUB_OUTPUT
 
       - name: using chill-viking/npm-ci
+        id: test-run
         uses: ./npm-ci # Use action from local folder
         with:
           working_directory: testing
@@ -130,12 +87,11 @@ jobs:
   post-cache-run:
     runs-on: ${{ inputs.runs_on }}
     needs:
-      - prepare-for-cache
       - pre-cache-run
-    name: "[post-cache]"
+    name: "cache run"
+    if: ${{ needs.pre-cache-run.outputs.used_cache != 'true' }}
     outputs:
       duration_ms: ${{ steps.timer-end.outputs.duration_ms }}
-      key_used: ${{ steps.cache-key.outputs.key }}
     steps:
       - name: "Use Node.js (${{ inputs.node_major }}.x)"
         uses: actions/setup-node@v3
@@ -145,7 +101,7 @@ jobs:
       - name: Download test folder
         uses: actions/download-artifact@v3
         with:
-          name: test_package_node_${{ inputs.node_major }}_${{ inputs.runs_on }}
+          name: test_package
           path: testing
 
       - name: Get Cache Key
@@ -204,18 +160,13 @@ jobs:
     needs:
       - pre-cache-run
       - post-cache-run
-    name: "[compare]"
+    name: "output results"
+    if: always()
     steps:
       - name: Compare times
+        if: ${{ needs.pre-cache-run.outputs.used_cache != 'true' }}
         shell: pwsh
         run: |
-          $preCacheKey = "${{ needs.pre-cache-run.outputs.key_used }}"
-          $postCacheKey = "${{ needs.post-cache-run.outputs.key_used }}"
-          if ($preCacheKey -ne $postCacheKey)
-          {
-              Write-Host "::error::pre-cache key [${preCacheKey}] was different to post-cache key [${postCacheKey}]"
-          }
-
           $preCacheDuration = [Float]::Parse("${{ needs.pre-cache-run.outputs.duration_ms }}")
           $postCacheDuration = [Float]::Parse("${{ needs.post-cache-run.outputs.duration_ms }}")
           $difference = $preCacheDuration - $postCacheDuration
@@ -235,3 +186,13 @@ jobs:
           Write-Output "| pre-cache | ${preCacheDuration}ms |" >> $env:GITHUB_STEP_SUMMARY
           Write-Output "| post-cache | ${postCacheDuration}ms |" >> $env:GITHUB_STEP_SUMMARY
           Write-Output "| difference | ${diffPercent}% ${diffTitle} |" >> $env:GITHUB_STEP_SUMMARY
+
+      - name: Compare times
+        if: ${{ needs.pre-cache-run.outputs.used_cache == 'true' }}
+        shell: pwsh
+        run: |
+          $preCacheDuration = [Float]::Parse("${{ needs.pre-cache-run.outputs.duration_ms }}")
+
+          Write-Output "## :white_check_mark: used cache on first attempt" >> $env:GITHUB_STEP_SUMMARY
+          Write-Output "" >> $env:GITHUB_STEP_SUMMARY
+          Write-Output "Completed in ${preCacheDuration}ms" >> $env:GITHUB_STEP_SUMMARY

--- a/.github/workflows/run-action.yml
+++ b/.github/workflows/run-action.yml
@@ -180,7 +180,7 @@ jobs:
             $initialTitle = ":warning:"
           }
 
-          Write-Output "## ${initialTitle} Times for node ${{ inputs.node_major }}.x on ${{ inputs.runs_on }}" >> $env:GITHUB_STEP_SUMMARY
+          Write-Output "## ${initialTitle} ${{ inputs.runs_on }} using node ${{ inputs.node_major }}.x" >> $env:GITHUB_STEP_SUMMARY
           Write-Output "| Step | Duration |" >> $env:GITHUB_STEP_SUMMARY
           Write-Output "| --- | --- |" >> $env:GITHUB_STEP_SUMMARY
           Write-Output "| pre-cache | ${preCacheDuration}ms |" >> $env:GITHUB_STEP_SUMMARY
@@ -193,6 +193,6 @@ jobs:
         run: |
           $preCacheDuration = [Float]::Parse("${{ needs.pre-cache-run.outputs.duration_ms }}")
 
-          Write-Output "## :white_check_mark: used cache on first attempt" >> $env:GITHUB_STEP_SUMMARY
+          Write-Output "## :white_check_mark:  ${{ inputs.runs_on }} using node ${{ inputs.node_major }}.x used cache" >> $env:GITHUB_STEP_SUMMARY
           Write-Output "" >> $env:GITHUB_STEP_SUMMARY
           Write-Output "Completed in ${preCacheDuration}ms" >> $env:GITHUB_STEP_SUMMARY

--- a/.github/workflows/run-action.yml
+++ b/.github/workflows/run-action.yml
@@ -36,7 +36,7 @@ jobs:
       - name: Download test folder
         uses: actions/download-artifact@v3
         with:
-          name: test_package
+          name: test_package_node_${{ inputs.node_major }}
           path: testing
 
       - name: Download action
@@ -101,7 +101,7 @@ jobs:
       - name: Download test folder
         uses: actions/download-artifact@v3
         with:
-          name: test_package
+          name: test_package_node_${{ inputs.node_major }}
           path: testing
 
       - name: Get Cache Key

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -26,6 +26,7 @@ jobs:
         uses: actions/checkout@v3
 
       - name: Cleanup
+        if: github.event_name == 'pull_request'
         run: |
           gh extension install actions/gh-actions-cache
 

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -26,7 +26,7 @@ jobs:
         uses: actions/checkout@v3
 
       - name: Cleanup
-        if: github.event_name == 'pull_request'
+        if: github.event_name != 'pull_request'
         run: |
           gh extension install actions/gh-actions-cache
 

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -47,8 +47,16 @@ jobs:
 
   create-package:
     runs-on: ubuntu-latest
-    name: init test package and upload action
+    name: init test package
+    strategy:
+      matrix:
+        node-version: [14, 16, 18]
     steps:
+      - name: "Use Node.js (${{ matrix.node-version }}.x)"
+        uses: actions/setup-node@v3
+        with:
+          node-version: ${{ matrix.node-version }}.x
+
       - name: Create test package
         run: |
           mkdir testing
@@ -60,9 +68,13 @@ jobs:
       - name: Upload testing as artifact
         uses: actions/upload-artifact@v3
         with:
-          name: test_package
+          name: test_package_node_${{ matrix.node-version }}
           path: testing
 
+  prepare-action:
+    runs-on: ubuntu-latest
+    name: upload action
+    steps:
       - name: Checkout latest version of action
         uses: actions/checkout@v3
         with:
@@ -77,6 +89,8 @@ jobs:
   test-action:
     needs:
       - create-package
+      - prepare-action
+      - clear-cache
     name: test action
     # create a matrix of node versions to test against
     strategy:

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -12,14 +12,39 @@ on: # yamllint disable-line rule:truthy
   schedule:
     - cron: "30 5,17 * * *"
 
-permissions:
-  contents: write
+permissions: write-all
 
 concurrency:
   group: ${{ github.ref }}-${{ github.workflow }}
   cancel-in-progress: true
 
 jobs:
+  clear-cache:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v3
+
+      - name: Cleanup
+        run: |
+          gh extension install actions/gh-actions-cache
+
+          REPO=${{ github.repository }}
+
+          echo "Fetching list of cache key"
+          cacheKeysForPR=$(gh actions-cache list -R $REPO | cut -f 1 )
+
+          ## Setting this to not fail the workflow while deleting cache keys. 
+          set +e
+          echo "Deleting caches..."
+          for cacheKey in $cacheKeysForPR
+          do
+              gh actions-cache delete $cacheKey -R $REPO --confirm
+          done
+          echo "Done"
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
   create-package:
     runs-on: ubuntu-latest
     name: init test package and upload action

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -26,7 +26,7 @@ jobs:
         uses: actions/checkout@v3
 
       - name: Cleanup
-        if: github.event_name != 'pull_request'
+        if: github.event_name == 'pull_request'
         run: |
           gh extension install actions/gh-actions-cache
 

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -29,7 +29,7 @@ jobs:
           mkdir testing
           cd testing
           npm init -y
-          npm install --save-dev @chill-viking/layout
+          npm install --save @chill-viking/layout
           rm -rf node_modules
 
       - name: Upload testing as artifact

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -35,7 +35,7 @@ jobs:
           echo "Fetching list of cache key"
           cacheKeysForPR=$(gh actions-cache list -R $REPO | cut -f 1 )
 
-          ## Setting this to not fail the workflow while deleting cache keys. 
+          ## Setting this to not fail the workflow while deleting cache keys.
           set +e
           echo "Deleting caches..."
           for cacheKey in $cacheKeysForPR

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -9,6 +9,7 @@ on: # yamllint disable-line rule:truthy
     branches: [main]
   pull_request:
     branches: [main]
+    types: [opened, synchronize, reopened, closed]
   schedule:
     - cron: "30 5,17 * * *"
 
@@ -49,6 +50,7 @@ jobs:
   create-package:
     runs-on: ubuntu-latest
     name: init test package
+    if: github.event.action != 'closed'
     strategy:
       matrix:
         node-version: [14, 16, 18]
@@ -75,6 +77,7 @@ jobs:
   prepare-action:
     runs-on: ubuntu-latest
     name: upload action
+    if: github.event.action != 'closed'
     steps:
       - name: Checkout latest version of action
         uses: actions/checkout@v3
@@ -93,6 +96,7 @@ jobs:
       - prepare-action
       - clear-cache
     name: test action
+    if: github.event.action != 'closed'
     # create a matrix of node versions to test against
     strategy:
       max-parallel: 5

--- a/.github/workflows/test-action.yml
+++ b/.github/workflows/test-action.yml
@@ -40,7 +40,7 @@ jobs:
           echo "Deleting caches..."
           for cacheKey in $cacheKeysForPR
           do
-              gh actions-cache delete $cacheKey -R $REPO --confirm
+            gh actions-cache delete "$cacheKey" -R $REPO --confirm
           done
           echo "Done"
         env:

--- a/README.md
+++ b/README.md
@@ -37,6 +37,12 @@ steps:
 |---------------------|-----------------------------------------------------------------------------|----------|---------|
 | `working_directory` | The directory to install dependencies in (location of `package-lock.json`). | No       | `./`    |
 
+## Outputs
+
+| Name                  | Description                                                             |
+|-----------------------|-------------------------------------------------------------------------|
+| `restored_from_cache` | `true` if the dependencies were restored from cache, `false` otherwise. |
+
 ## Compatibility
 
 The following combinations of runner and Node.js version(s) are checked in the workflow [![Test chill-viking/npm-ci](https://github.com/chill-viking/npm-ci/actions/workflows/test-action.yml/badge.svg)](https://github.com/chill-viking/npm-ci/actions/workflows/test-action.yml)

--- a/action.yml
+++ b/action.yml
@@ -14,6 +14,11 @@ inputs:
     description: "Working directory to use"
     default: "./"
 
+outputs:
+  restored_from_cache:
+    description: "True if dependencies were restored from cache"
+    value: ${{ steps.cache-node.outputs.cache-hit }}
+
 runs:
   using: "composite"
   steps:

